File: src/services/ops_intel_collector.py
import asyncio
import json
import re
from datetime import datetime
from html import unescape
from urllib.parse import urljoin, urlparse

import httpx

from ..config import BASE_DIR, settings
from ..database import SessionLocal
from ..models import OpsIntelItem, OpsIntelRun
from .official_event_parsers import collect_recent_official_events

_task: asyncio.Task | None = None
OFFICIAL_HOSTS = ["faa.gov", "ntsb.gov", "nasa.gov", "icao.int", "easa.europa.eu"]
EVENT_KEYWORDS = [
    "accident", "incident", "investigation", "safety", "recommendation", "safo", "info",
    "advisory", "airworthiness", "directive", "etops", "diversion", "fatigue", "crm",
    "runway", "unstable", "engine", "fire", "smoke", "gps", "gnss", "jamming", "spoofing",
    "mel", "dispatch", "training", "part-121", "part 121", "part-135", "part 135",
]
REPORT_KEYWORDS = [
    "report", "final", "preliminary", "investigation", "recommendation", "safo", "info",
    "advisory circular", "airworthiness directive", "directive", "safety alert",
    "accident", "incident", "asrs", "callback", "lessons learned",
]
SKIP_LABELS = ["skip to", "main content", "enable javascript", "subscribe", "login", "sign in", "privacy", "contact"]
MAX_EVENT_DETAIL_FETCHES_PER_SOURCE = 4


def load_sources() -> list[dict]:
    with open(BASE_DIR / "data" / "sources.json", encoding="utf-8") as f:
        return json.load(f)


def html_title(text: str) -> str:
    match = re.search(r"<title[^>]*>(.*?)</title>", text, re.IGNORECASE | re.DOTALL)
    if not match:
        return ""
    return re.sub(r"\s+", " ", match.group(1)).strip()


def clean_text(text: str) -> str:
    return re.sub(r"\s+", " ", unescape(re.sub(r"<[^>]+>", " ", text))).strip()


def extract_main_text(html: str, limit: int = 1400) -> str:
    text = re.sub(r"(?is)<(script|style|nav|footer|header).*?>.*?</\1>", " ", html)
    text = clean_text(text)
    return text[:limit]


def is_same_site(base_url: str, url: str) -> bool:
    return urlparse(base_url).netloc == urlparse(url).netloc


def is_official_url(url: str) -> bool:
    host = urlparse(url).netloc.lower()
    return any(host.endswith(domain) for domain in OFFICIAL_HOSTS)


def classify_title(title: str, source: dict) -> tuple[str, str]:
    text = title.lower()
    if any(word in text for word in ["accident", "incident", "investigation", "runway", "engine", "fire", "smoke"]):
        return "Accident / Incident", "Medium"
    if any(word in text for word in ["regulation", "advisory", "directive", "safo", "info", "airworthiness"]):
        return "Regulation", "Medium"
    if any(word in text for word in ["fatigue", "crm", "human factors"]):
        return "Human Factors / CRM", "Medium"
    if any(word in text for word in ["gps", "gnss", "jamming", "spoofing", "security"]):
        return "Security / External Threat", "High"
    if any(word in text for word in ["etops", "dispatch", "diversion", "fuel", "mel"]):
        return "Flight Operations", "Medium"
    return source["category"], "Low"


def quality_status(source: dict, status_code: int) -> str:
    if status_code >= 400:
        return "needs_source_review"
    if is_official_url(source["url"]):
        return "official_source"
    return "supplementary_source"


def candidate_quality(source: dict, url: str, title: str, text: str, status_code: int) -> str:
    if status_code >= 400:
        return "needs_source_review"
    if not is_official_url(url):
        return "supplementary_source"
    haystack = f"{title} {url} {text[:800]}".lower()
    if any(keyword in haystack for keyword in REPORT_KEYWORDS):
        return "official_report_candidate"
    return "official_source"


def detail_summary(source: dict, title: str, text: str, status_code: int) -> str:
    status = quality_status(source, status_code)
    if not text:
        return f"Collected event candidate from {source['name']}: {title}. Detail page text was not available. Quality status: {status}."
    return f"{title}. Source: {source['name']}. Quality status: {status}. Extract: {text}"


def detail_operational_lesson(category: str, title: str) -> str:
    if category == "Accident / Incident":
        return f"Use this item to brief event precursors, crew decision gates, and procedural defenses. Do not infer cause or fault until official findings confirm them."
    if category == "Regulation":
        return f"Check whether this item changes SOP, training, dispatch release, maintenance control, or compliance monitoring."
    if category == "Security / External Threat":
        return f"Review threat recognition, reporting, navigation cross-checks, and dispatch coordination for routes exposed to this threat."
    return f"Review this item for practical briefing value, training implications, and long-haul widebody operational relevance."


def extract_event_links(html: str, source: dict, limit: int = 8) -> list[dict]:
    links = []
    seen = set()
    for match in re.finditer(r'<a\s+[^>]*href=["\']([^"\']+)["\'][^>]*>(.*?)</a>', html, re.IGNORECASE | re.DOTALL):
        href, label_html = match.groups()
        url = urljoin(source["url"], href)
        if not url.startswith("http") or not is_same_site(source["url"], url) or url in seen:
            continue
        label = clean_text(label_html)
        haystack = f"{label} {url}".lower()
        if (
            len(label) < 8
            or any(skip in haystack for skip in SKIP_LABELS)
            or not any(keyword in haystack for keyword in EVENT_KEYWORDS)
        ):
            continue
        category, severity = classify_title(label, source)
        links.append({"url": url, "title": label[:240], "category": category, "severity": severity})
        seen.add(url)
        if len(links) >= limit:
            break
    return links


def item_summary(source: dict, status_code: int, title: str) -> str:
    return f"Periodic source check saved for {source['name']}. HTTP status {status_code}. Latest page title: {title or 'not available'}."


def operational_lesson(source: dict) -> str:
    return f"Review new {source['category']} material from {source['name']} for procedure, training, dispatch, or safety-management relevance before using it in pilot briefings."


def recommended_action(source: dict) -> str:
    if source["category"] == "Regulation":
        return "Screen for regulatory or procedural changes and assign SOP/training review if applicable."
    if source["category"] == "Accident / Incident":
        return "Check for new factual reports, classify severity, and extract operational lessons without assigning blame."
    if source["category"] == "Human Factors / CRM":
        return "Review for recurrent training, fatigue, CRM, and decision-making discussion points."
    return "Review source updates and convert relevant items into the operational lessons database."


def upsert_item(db, source: dict, url: str, title: str, category: str, severity: str, status_code: int, summary: str, extra_tags: list[str] | None = None) -> bool:
    item = db.query(OpsIntelItem).filter(OpsIntelItem.source_url == url).first()
    created = False
    if not item:
        item = OpsIntelItem(source_name=source["name"], source_url=url)
        db.add(item)
        created = True
    item.title = title or source["name"]
    item.category = category
    item.severity = severity
    item.summary = summary
    item.operational_lesson = operational_lesson({"name": source["name"], "category": category})
    item.a350_b787_applicability = "Screen for relevance to A350/B787 long-haul operations, especially ETOPS, fatigue, dispatch, MEL, navigation, and approach threats."
    item.recommended_action = recommended_action({"category": category})
    item.tags = json.dumps(list(dict.fromkeys([*source.get("tags", []), category, severity, *(extra_tags or [])])))
    item.last_status = status_code
    item.last_checked_at = datetime.utcnow()
    item.updated_at = datetime.utcnow()
    return created


async def fetch_source(client: httpx.AsyncClient, source: dict) -> dict:
    response = await client.get(source["url"])
    links = extract_event_links(response.text, source)
    enriched_links = []
    for link in links[:MAX_EVENT_DETAIL_FETCHES_PER_SOURCE]:
        try:
            detail_response = await client.get(link["url"])
            text = extract_main_text(detail_response.text)
            category, severity = classify_title(f"{link['title']} {text[:500]}", source)
            enriched_links.append({
                **link,
                "category": category,
                "severity": severity,
                "detail_status": detail_response.status_code,
                "detail_text": text,
            })
        except Exception:
            enriched_links.append({**link, "detail_status": response.status_code, "detail_text": ""})
    enriched_links.extend(links[MAX_EVENT_DETAIL_FETCHES_PER_SOURCE:])
    return {
        "source": source,
        "status_code": response.status_code,
        "title": html_title(response.text),
        "links": enriched_links,
    }


async def collect_once() -> dict:
    sources = load_sources()
    db = SessionLocal()
    run = OpsIntelRun()
    db.add(run)
    db.commit()
    saved = 0
    seen_urls = set()
    try:
        async with httpx.AsyncClient(timeout=12, follow_redirects=True, headers={"User-Agent": "OpsBriefing/0.1"}) as client:
            results = await asyncio.gather(*(fetch_source(client, source) for source in sources), return_exceptions=True)
        for result in results:
            if isinstance(result, Exception):
                continue
            source = result["source"]
            if source["url"] in seen_urls:
                continue
            seen_urls.add(source["url"])
            if upsert_item(
                db,
                source,
                source["url"],
                result["title"] or source["name"],
                source["category"],
                "Low",
                result["status_code"],
                item_summary(source, result["status_code"], result["title"]),
            ):
                saved += 1
            for link in result["links"]:
                if link["url"] in seen_urls:
                    continue
                seen_urls.add(link["url"])
                summary = f"Collected event candidate from {source['name']}: {link['title']}."
                quality = candidate_quality(source, link["url"], link["title"], link.get("detail_text", ""), link.get("detail_status", result["status_code"]))
                if "detail_text" in link:
                    summary = detail_summary(source, link["title"], link.get("detail_text", ""), link.get("detail_status", result["status_code"]))
                if upsert_item(
                    db,
                    source,
                    link["url"],
                    link["title"],
                    link["category"],
                    link["severity"],
                    link.get("detail_status", result["status_code"]),
                    summary,
                    [quality],
                ):
                    saved += 1
                item = db.query(OpsIntelItem).filter(OpsIntelItem.source_url == link["url"]).first()
                if item and "detail_text" in link:
                    item.operational_lesson = detail_operational_lesson(link["category"], link["title"])
                    tags = json.loads(item.tags or "[]")
                    tags.append(quality)
                    item.tags = json.dumps(list(dict.fromkeys(tags)))
        db.commit()
        official_result = await collect_recent_official_events()
        saved += official_result.get("items_saved", 0)
        run.status = "complete"
        run.items_checked = len(sources) + official_result.get("items_checked", 0)
        run.items_saved = saved
        run.finished_at = datetime.utcnow()
        db.commit()
        return {
            "status": run.status,
            "items_checked": run.items_checked,
            "items_saved": run.items_saved,
            "official_recent": official_result,
        }
    except Exception as exc:
        db.rollback()
        run = db.get(OpsIntelRun, run.id)
        run.status = "failed"
        run.error = str(exc)
        run.finished_at = datetime.utcnow()
        db.commit()
        return {"status": run.status, "error": run.error}
    finally:
        db.close()


async def periodic_collect_loop():
    while True:
        await collect_once()
        await asyncio.sleep(max(1, settings.ops_intel_interval_hours) * 3600)


def start_periodic_collector():
    global _task
    if settings.ops_intel_autostart and (_task is None or _task.done()):
        _task = asyncio.create_task(periodic_collect_loop())


def stop_periodic_collector():
    global _task
    if _task and not _task.done():
        _task.cancel()
    _task = None


def refine_official_items() -> dict:
    db = SessionLocal()
    counts = {"official_report_candidate": 0, "official_source": 0, "supplementary_source": 0, "needs_source_review": 0}
    try:
        rows = db.query(OpsIntelItem).all()
        for item in rows:
            quality = candidate_quality(
                {"url": item.source_url},
                item.source_url,
                item.title or "",
                item.summary or "",
                item.last_status or 0,
            )
            counts[quality] += 1
            tags = json.loads(item.tags or "[]")
            tags = [tag for tag in tags if tag not in counts]
            tags.append(quality)
            item.tags = json.dumps(list(dict.fromkeys(tags)))
            if quality == "official_report_candidate" and item.severity == "Low":
                item.severity = "Medium"
            item.updated_at = datetime.utcnow()
        db.commit()
        return {"items_refined": len(rows), **counts}
    finally:
        db.close()

File: worker/src/services/ops_intel_collector.ts
import type { OpsIntelItemRow, Source, Env } from "../types";
import { SOURCES } from "../data/sources";
import { collectRecentOfficialEvents } from "./official_event_parsers";
import { enrichWithLLM } from "./llm_classifier";

const OFFICIAL_HOSTS = ["faa.gov", "ntsb.gov", "nasa.gov", "icao.int", "easa.europa.eu", "skybrary.aero"];
const EVENT_KEYWORDS = ["accident","incident","investigation","safety","recommendation","safo","advisory","airworthiness","directive","runway","engine","fire","smoke","gps","gnss","jamming","mel","training","part-121","part 135","approach","landing","departure","takeoff"];
const REPORT_KEYWORDS = ["report","final","preliminary","investigation","recommendation","safo","advisory circular","airworthiness directive","safety alert","accident","incident","asrs","callback","lessons learned"];
const SKIP_LABELS = ["skip to","main content","enable javascript","subscribe","login","sign in","privacy","contact"];
const DEFAULT_MAX_DETAIL_FETCHES = 8;

function htmlTitle(text: string): string {
  const m = text.match(/<title[^>]*>([\s\S]*?)<\/title>/i);
  return m ? m[1].replace(/\s+/g, " ").trim() : "";
}

function cleanText(text: string): string {
  return text.replace(/<[^>]+>/g, " ").replace(/\s+/g, " ").trim();
}

function extractMainText(html: string, limit = 1400): string {
  const stripped = html.replace(/<(script|style|nav|footer|header)[\s\S]*?<\/\1>/gi, " ");
  return cleanText(stripped).slice(0, limit);
}

function isOfficialUrl(url: string): boolean {
  try { return OFFICIAL_HOSTS.some(h => new URL(url).hostname.endsWith(h)); } catch { return false; }
}

function isSameSite(base: string, url: string): boolean {
  try { return new URL(base).hostname === new URL(url).hostname; } catch { return false; }
}

function classifyTitle(title: string, source: Source): [string, string] {
  const t = title.toLowerCase();
  if (["accident","incident","investigation","runway","engine","fire","smoke"].some(w => t.includes(w))) return ["Accident / Incident", "Medium"];
  if (["regulation","advisory","directive","safo","airworthiness"].some(w => t.includes(w))) return ["Regulation", "Medium"];
  if (["fatigue","crm","human factors"].some(w => t.includes(w))) return ["Human Factors / CRM", "Medium"];
  if (["gps","gnss","jamming","spoofing","security"].some(w => t.includes(w))) return ["Security / External Threat", "High"];
  return [source.category, "Low"];
}

function candidateQuality(sourceUrl: string, url: string, title: string, text: string, statusCode: number): string {
  if (statusCode >= 400) return "needs_source_review";
  if (!isOfficialUrl(url)) return "supplementary_source";
  const haystack = `${title} ${url} ${text.slice(0, 800)}`.toLowerCase();
  return REPORT_KEYWORDS.some(k => haystack.includes(k)) ? "official_report_candidate" : "official_source";
}

function extractEventLinks(html: string, source: Source): { url: string; title: string; category: string; severity: string }[] {
  const links: { url: string; title: string; category: string; severity: string }[] = [];
  const seen = new Set<string>();
  for (const m of html.matchAll(/<a\s+[^>]*href=["']([^"']+)["'][^>]*>([\s\S]*?)<\/a>/gi)) {
    const [, href, labelHtml] = m;
    let url: string;
    try { url = new URL(href, source.url).href; } catch { continue; }
    if (!url.startsWith("http") || !isSameSite(source.url, url) || seen.has(url)) continue;
    const label = cleanText(labelHtml);
    const haystack = `${label} ${url}`.toLowerCase();
    if (label.length < 8 || SKIP_LABELS.some(s => haystack.includes(s)) || !EVENT_KEYWORDS.some(k => haystack.includes(k))) continue;
    const [category, severity] = classifyTitle(label, source);
    links.push({ url, title: label.slice(0, 240), category, severity });
    seen.add(url);
    if (links.length >= 8) break;
  }
  return links;
}

async function upsertItem(db: D1Database, source: Source, url: string, title: string, category: string, severity: string, statusCode: number, summary: string, extraTags: string[] = []): Promise<boolean> {
  const existing = await db.prepare("SELECT id FROM ops_intel_items WHERE source_url = ?").bind(url).first<{ id: number }>();
  const tags = JSON.stringify([...new Set([...source.tags, category, severity, ...extraTags])]);
  const now = new Date().toISOString();
  const lesson = `Review new ${category} material from ${source.name} for procedure, training, dispatch, or safety-management relevance before using it in pilot briefings.`;
  const a350 = "Screen for relevance to A350/B787 long-haul operations, especially ETOPS, fatigue, dispatch, MEL, navigation, and approach threats.";
  const action = category === "Regulation" ? "Screen for regulatory or procedural changes and assign SOP/training review if applicable."
    : category === "Accident / Incident" ? "Check for new factual reports, classify severity, and extract operational lessons without assigning blame."
    : "Review source updates and convert relevant items into the operational lessons database.";

  if (!existing) {
    await db.prepare("INSERT INTO ops_intel_items (source_name,source_url,title,category,severity,summary,operational_lesson,a350_b787_applicability,recommended_action,tags,last_status,last_checked_at,created_at,updated_at) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?)")
      .bind(source.name, url, title || source.name, category, severity, summary, lesson, a350, action, tags, statusCode, now, now, now).run();
    return true;
  }
  await db.prepare("UPDATE ops_intel_items SET title=?,category=?,severity=?,summary=?,operational_lesson=?,a350_b787_applicability=?,recommended_action=?,tags=?,last_status=?,last_checked_at=?,updated_at=? WHERE source_url=?")
    .bind(title || source.name, category, severity, summary, lesson, a350, action, tags, statusCode, now, now, url).run();
  return false;
}

async function fetchSource(source: Source, maxDetailFetches: number, priorityHosts: Set<string>): Promise<{ source: Source; statusCode: number; title: string; links: { url: string; title: string; category: string; severity: string; detailStatus: number; detailText: string }[] }> {
  const res = await fetch(source.url, { headers: { "User-Agent": "OpsBriefing/0.1" }, signal: AbortSignal.timeout(12000), redirect: "follow" });
  const html = await res.text();
  const rawLinks = extractEventLinks(html, source);

  // 우선 소스는 더 많은 링크 상세 탐색
  const isPriority = priorityHosts.size > 0 && [...priorityHosts].some(h => source.url.includes(h));
  const detailLimit = isPriority ? Math.max(maxDetailFetches, 12) : maxDetailFetches;

  const enriched = [];
  for (const link of rawLinks.slice(0, detailLimit)) {
    try {
      const dr = await fetch(link.url, { headers: { "User-Agent": "OpsBriefing/0.1" }, signal: AbortSignal.timeout(12000), redirect: "follow" });
      const dt = extractMainText(await dr.text());
      const [cat, sev] = classifyTitle(`${link.title} ${dt.slice(0, 500)}`, source);
      enriched.push({ ...link, category: cat, severity: sev, detailStatus: dr.status, detailText: dt });
    } catch {
      enriched.push({ ...link, detailStatus: res.status, detailText: "" });
    }
  }
  enriched.push(...rawLinks.slice(detailLimit).map(l => ({ ...l, detailStatus: res.status, detailText: "" })));
  return { source, statusCode: res.status, title: htmlTitle(html), links: enriched };
}

export async function collectOnce(db: D1Database, env?: Env): Promise<Record<string, unknown>> {
  const runInsert = await db.prepare("INSERT INTO ops_intel_runs (started_at,status) VALUES (?,?) RETURNING id").bind(new Date().toISOString(), "running").first<{ id: number }>();
  const runId = runInsert!.id;

  const maxDetailFetches = parseInt(env?.MAX_DETAIL_FETCHES ?? String(DEFAULT_MAX_DETAIL_FETCHES), 10);
  const priorityHosts = new Set((env?.PRIORITY_FULL_SCAN ?? "faa.gov,ntsb.gov,easa.europa.eu,icao.int").split(",").map(s => s.trim()));

  let saved = 0;
  const seenUrls = new Set<string>();

  try {
    const results = await Promise.allSettled(SOURCES.map(s => fetchSource(s, maxDetailFetches, priorityHosts)));
    for (const result of results) {
      if (result.status === "rejected") continue;
      const { source, statusCode, title, links } = result.value;
      if (!seenUrls.has(source.url)) {
        seenUrls.add(source.url);
        if (await upsertItem(db, source, source.url, title || source.name, source.category, "Low", statusCode,
            `Periodic source check saved for ${source.name}. HTTP status ${statusCode}. Latest page title: ${title || "not available"}.`)) saved++;
      }
      for (const link of links) {
        if (seenUrls.has(link.url)) continue;
        seenUrls.add(link.url);
        const quality = candidateQuality(source.url, link.url, link.title, link.detailText, link.detailStatus);
        const summary = link.detailText
          ? `${link.title}. Source: ${source.name}. Quality: ${quality}. Extract: ${link.detailText}`
          : `Collected event candidate from ${source.name}: ${link.title}.`;
        if (await upsertItem(db, source, link.url, link.title, link.category, link.severity, link.detailStatus, summary, [quality])) saved++;
      }
    }

    const officialResult = await collectRecentOfficialEvents(db, 20);
    saved += (officialResult.items_saved as number) ?? 0;

    // LLM enrichment: Workers AI가 있으면 신규 항목 일괄 분류
    let llmResult: Record<string, unknown> = { skipped: "no AI binding" };
    if (env?.AI) {
      try {
        llmResult = await enrichWithLLM(env.AI, db, 20);
      } catch (e) {
        llmResult = { error: String(e) };
      }
    }

    const now = new Date().toISOString();
    await db.prepare("UPDATE ops_intel_runs SET status=?,items_checked=?,items_saved=?,finished_at=? WHERE id=?")
      .bind("complete", SOURCES.length, saved, now, runId).run();
    return { status: "complete", items_checked: SOURCES.length, items_saved: saved, official_recent: officialResult, llm_enrichment: llmResult };
  } catch (err) {
    const now = new Date().toISOString();
    await db.prepare("UPDATE ops_intel_runs SET status=?,error=?,finished_at=? WHERE id=?")
      .bind("failed", String(err), now, runId).run();
    return { status: "failed", error: String(err) };
  }
}

export async function refineOfficialItems(db: D1Database): Promise<Record<string, unknown>> {
  const { results } = await db.prepare("SELECT * FROM ops_intel_items").all<OpsIntelItemRow>();
  const counts: Record<string, number> = { official_report_candidate: 0, official_source: 0, supplementary_source: 0, needs_source_review: 0 };
  for (const item of results) {
    const quality = candidateQuality(item.source_url, item.source_url, item.title ?? "", item.summary ?? "", item.last_status ?? 0);
    counts[quality] = (counts[quality] ?? 0) + 1;
    const tags = [...new Set([...JSON.parse(item.tags ?? "[]") as string[], quality])];
    const now = new Date().toISOString();
    await db.prepare("UPDATE ops_intel_items SET tags=?,severity=?,updated_at=? WHERE source_url=?")
      .bind(JSON.stringify(tags), quality === "official_report_candidate" && item.severity === "Low" ? "Medium" : item.severity, now, item.source_url).run();
  }
  return { items_refined: results.length, ...counts };
}
